compiler/rustc_ast/src/ast.rs

```diff
@@ -3374,13 +3374,19 @@ pub struct ConstItem {
     pub expr: Option<P<Expr>>,
 }
 
+#[derive(PartialEq, Eq, Clone, Copy, Encodable, Decodable, Debug, HashStable_Generic)]
+pub enum ExternCrateKind {
+    Stable,
+    Default,
+}
+
 // Adding a new variant? Please update `test_item` in `tests/ui/macros/stringify.rs`.
 #[derive(Clone, Encodable, Decodable, Debug)]
 pub enum ItemKind {
     /// An `extern crate` item, with the optional *original* crate name if the crate was renamed.
     ///
     /// E.g., `extern crate foo` or `extern crate foo_bar as foo`.
-    ExternCrate(Option<Symbol>),
+    ExternCrate(ExternCrateKind, Option<Symbol>),
```

Comment:
What do `ItemKind::ExternCrate(ExternCrateKind::Stable, _)` and `ItemKind::ExternCrate(ExternCrateKind::Default, _)` correlate to? Are you sure these should be AST items?

Reply:
@Bryanskiy Is this supposed to distinguish `extern dyn crate`? Shouldn't that be its own `ItemKind`?

Reply:
Yeah, agreed that these should likely be different items unless their semantics are basically the same.

Reply:
@compiler-errors In particular, crates introduced into view by `extern dyn crate` are only supposed to have items marked with `#[export]` be considered "reachable" by `use`. It's sort of like introducing a new `pub(dyn)` level of visibility.

Reply:
I'd suggest using an attribute for now, to keep the feature more isolated and avoid touching code that isn't directly related to it (by not modifying AST).

Reply:
@petrochenkov Wouldn't that introduce the risk that code handling `ItemKind::ExternCrate` doesn't check for the attribute, and thus doesn't handle the name resolution differences properly?